modules/rostests/apitests/compiler/pseh.c

```diff
@@ -2765,8 +2771,160 @@ struct subtest
+        okx(EH3Registration->ExceptionHandler != NULL);
+#ifdef _MSC_VER
+        okx(EH3Registration->ExceptionHandler == _except_handler3);
+#endif
+        okx(EH3Registration->ScopeTable != NULL);
+        okx(EH3Registration->TryLevel == 0);
+        ScopeTable = EH3Registration->ScopeTable;
+        okx(ScopeTable->EnclosingLevel == 0xFFFFFFFF);
+        okx(ScopeTable->FilterFunc != NULL);
+        okx(ScopeTable->HandlerFunc != NULL);
+    }
+    _SEH2_EXCEPT(1)
+    {
+    }
+    _SEH2_END;
+}
+
+DECLSPEC_NOINLINE
+static void Test_structs_seh_finally(void)
+{
+    PEH3_EXCEPTION_REGISTRATION EH3Registration;
+    PSCOPETABLE_ENTRY ScopeTable;
+
+    _SEH2_TRY
+    {
+        EH3Registration = (PEH3_EXCEPTION_REGISTRATION)__readfsdword(0);
+        okx(EH3Registration != NULL);
+        okx(EH3Registration != (PVOID)0xFFFFFFFF);
+        okx(EH3Registration->Next != (PVOID)0xFFFFFFFF);
+        okx(EH3Registration->ExceptionHandler != NULL);
+#ifdef _MSC_VER
+        okx(EH3Registration->ExceptionHandler == _except_handler3);
+#endif
+        okx(EH3Registration->ScopeTable != NULL);
+        okx(EH3Registration->TryLevel == 0);
+        ScopeTable = EH3Registration->ScopeTable;
+        okx(ScopeTable->EnclosingLevel == 0xFFFFFFFF);
+        todo_pseh okx(ScopeTable->FilterFunc == NULL);
+        todo_pseh okx(ScopeTable->HandlerFunc != NULL);
+    }
+    _SEH2_FINALLY
+    {
+    }
+    _SEH2_END;
+}
+
+DECLSPEC_NOINLINE
+static void Test_structs_seh_nested(void)
+{
+    PEH3_EXCEPTION_REGISTRATION EH3Registration1, EH3Registration2;
+    PSCOPETABLE_ENTRY ScopeTable;
+
+    _SEH2_TRY
+    {
+        EH3Registration1 = (PEH3_EXCEPTION_REGISTRATION)__readfsdword(0);
+        okx(EH3Registration1 != NULL);
+        okx(EH3Registration1 != (PVOID)0xFFFFFFFF);
+        okx(EH3Registration1->TryLevel == 0);
+        _SEH2_TRY
+        {
+            EH3Registration2 = (PEH3_EXCEPTION_REGISTRATION)__readfsdword(0);
+            okx(EH3Registration2 != NULL);
+            okx(EH3Registration2 != (PVOID)0xFFFFFFFF);
+            okx(EH3Registration2 == EH3Registration1);
+            okx(EH3Registration2->ExceptionHandler != NULL);
+#ifdef _MSC_VER
+            okx(EH3Registration2->ExceptionHandler == _except_handler3);
+#endif
+            okx(EH3Registration2->ScopeTable != NULL);
+            okx(EH3Registration2->ScopeTable == EH3Registration1->ScopeTable);
+            todo_pseh okx(EH3Registration2->TryLevel == 1);
+            ScopeTable = EH3Registration2->ScopeTable;
+            okx(ScopeTable[0].EnclosingLevel == 0xFFFFFFFF);
+            okx(ScopeTable[0].FilterFunc != NULL);
+            okx(ScopeTable[0].HandlerFunc != NULL);
+            okx(ScopeTable[1].EnclosingLevel == 0);
+            todo_pseh okx(ScopeTable[1].FilterFunc == NULL);
+            todo_pseh okx(ScopeTable[1].HandlerFunc != NULL);
+        }
+        _SEH2_FINALLY
+        {
+        }
+        _SEH2_END;
+    }
```

Comment:
Would it make sense to also add some validation tests at the end of the top-level TRY block, after this nested TRY/FINALLY?